ractor_actors/examples/tcp.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Sean Lawlor
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree.
+
+extern crate ractor_actors;
+
+use chrono::{DateTime, Utc};
+use ractor::{Actor, ActorProcessingErr, ActorRef};
+use ractor_actors::net::tcp::listener::*;
+use ractor_actors::net::tcp::session::*;
+use ractor_actors::net::tcp::*;
+use ractor_actors::watchdog;
+use ractor_actors::watchdog::TimeoutStrategy;
+use std::error::Error;
+use std::str::FromStr;
+use std::time::SystemTime;
+
+struct MyServer;
+
+struct MyServerState {}
+struct MyServerArgs {
+    /// Controls if the watchdog is used
+    watchdog: bool,
+
+    /// The port to listen on
+    port: NetworkPort,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl Actor for MyServer {
+    type Msg = ();
+    type State = MyServerState;
+    type Arguments = MyServerArgs;
+
+    async fn pre_start(
+        &self,
+        myself: ActorRef<Self::Msg>,
+        MyServerArgs { watchdog, port }: Self::Arguments,
+    ) -> Result<Self::State, ActorProcessingErr> {
+        let acceptor = MyServerSocketAcceptor { watchdog };
+
+        let _ = myself
+            .spawn_linked(
+                Some(format!("listener-{}", port)),
+                Listener::new(),
+                ListenerStartupArgs {
+                    port,
+                    encryption: IncomingEncryptionMode::Raw,
+                    acceptor,
+                },
+            )
+            .await?;
+
+        Ok(Self::State {})
+    }
+}
+
+struct MyServerSocketAcceptor {
+    watchdog: bool,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl SessionAcceptor for MyServerSocketAcceptor {
+    async fn new_session(&self, stream: NetworkStream) -> Result<(), ActorProcessingErr> {
+        tracing::info!("New connection: {}", stream.peer_addr());
+        Actor::spawn(
+            Some(format!("MySession-{}", stream.peer_addr().port())),
+            MySession {
+                info: stream.info(),
+            },
+            MySessionArgs {
+                watchdog: self.watchdog,
+                stream,
+            },
+        )
+        .await
+        .map_err(|e| ActorProcessingErr::from(e))?;
+
+        Ok(())
+    }
+}
+
+struct MySession {
+    info: NetworkStreamInfo,
+}
+
+enum MySessionMsg {
+    FrameReady(Frame),
+}
+
+struct MySessionState {
+    watchdog: bool,
+    session: ActorRef<TcpSessionMessage>,
+}
+
+struct MySessionArgs {
+    watchdog: bool,
+    stream: NetworkStream,
+}
+
+struct MyFrameReceiver {
+    session: ActorRef<MySessionMsg>,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl FrameReceiver for MyFrameReceiver {
+    async fn frame_ready(&self, f: Frame) -> Result<(), ActorProcessingErr> {
```

Comment:
in retrospect it probably makes sense for this to take `&mut self` and have an owned copy for each `TcpSession`. That way it can act as the "state" of the tcp session and you can put whatever logic here for every received frame.

Additionally maybe it should receive the TcpSession handle as an argument, so if it needs to reply, it can.

Reply:
That might be a good idea, but I think I have to try it out a bit first.

If I look at this from an Erlang point of view, it feels a little bit strange that the behavior isn't inside an actor. The FrameReceiver has no way of getting messages and is quite limited in what it can do. I think you would end up with a applicaion-level session actor anyway quite fast.